cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/remote/RemoteLogAppender.java

```diff
@@ -0,0 +1,174 @@
+    this.remoteClient = new RemoteClient(discoveryServiceClient, Constants.Service.LOG_BUFFER_SERVICE,
+                                         new DefaultHttpRequestConfig(), "/v1/");
+  }
+
+  @Override
+  public void start() {
+    publisher.startAndWait();
+    addInfo("Successfully started " + APPENDER_NAME);
+    super.start();
+  }
+
+  @Override
+  public void stop() {
+    publisher.stopAndWait();
+    addInfo("Successfully stopped " + APPENDER_NAME);
+    super.stop();
+  }
+
+  @Override
+  protected void appendEvent(LogMessage logMessage) {
+    logMessage.prepareForDeferredProcessing();
+    logMessage.getCallerData();
+
+    try {
+      publisher.addMessage(logMessage);
+    } catch (InterruptedException e) {
+      addInfo("Interrupted when adding log message to queue: " + logMessage.getFormattedMessage());
+    }
+  }
+
+  /**
+   * Publisher service to publish logs to log saver.
+   */
+  private final class RemoteLogPublisher extends AbstractLogPublisher<Map.Entry<Integer, byte[]>> {
+
+    private final int numPartitions;
+    private final LoggingEventSerializer loggingEventSerializer;
+    private final LogPartitionType logPartitionType;
+    private final DatumWriter<GenericRecord> datumWriter;
+
+    private RemoteLogPublisher(CConfiguration cConf) {
+      super(cConf.getInt(Constants.Logging.APPENDER_QUEUE_SIZE, 512),
+            RetryStrategies.fromConfiguration(cConf, "system.log.process."));
+      this.numPartitions = cConf.getInt(Constants.Logging.NUM_PARTITIONS);
+      this.loggingEventSerializer = new LoggingEventSerializer();
+      this.logPartitionType =
+        LogPartitionType.valueOf(cConf.get(Constants.Logging.LOG_PUBLISH_PARTITION_KEY).toUpperCase());
+      datumWriter = new GenericDatumWriter<>(LogSchema.LogBufferRequest.SCHEMA);
+    }
+
+    @Override
+    protected Map.Entry<Integer, byte[]> createMessage(LogMessage logMessage) {
+      String partitionKey = logPartitionType.getPartitionKey(logMessage.getLoggingContext());
+      int partition = partition(partitionKey, numPartitions);
+      return new AbstractMap.SimpleEntry<>(partition, loggingEventSerializer.toBytes(logMessage));
+    }
+
+    @Override
+    protected void publish(List<Map.Entry<Integer, byte[]>> logMessages) throws Exception {
+      // Group the log messages by partition and then publish all messages to their respective partitions
+      Map<Integer, List<byte[]>> partitionedMessages = new HashMap<>();
+      for (Map.Entry<Integer, byte[]> logMessage : logMessages) {
+        List<byte[]> messages = partitionedMessages.computeIfAbsent(logMessage.getKey(), k -> new ArrayList<>());
+        messages.add(logMessage.getValue());
+      }
+
+      try (ByteArrayOutputStream os = new ByteArrayOutputStream()) {
+        for (Map.Entry<Integer, List<byte[]>> partition : partitionedMessages.entrySet()) {
+          encodeEvents(os, datumWriter, partition.getKey(), partition.getValue());
```

Comment:
The schema is not good. The partition should be part of the request path and the request body should just be an avro array.

Reply:
Yes that should be addressed as part of https://github.com/cdapio/cdap/pull/11094#discussion_r259196628 I was not aware of how routing will be used in future.